[#14] LiveData 를 Flow 로 변경

fried-shoes opened the pull request:
## 2025.09.24. Issue #14 에 대한 PR

### 1. `LiveData` 를 `Flow` 로 변경하기
- `LiveData`, `MutableLiveData` 를 각각 `StateFlow`, `MutableStateFlow` 로 변경
- UI 의 `Lifecycle` 이 `STARTED` 일 때 시작하도록 `repeatOnLifecycle(Lifecycle.State.STARTED)` 설정

### 2. 추가 변경사항
- 당초 `MutableStateFlow` 의 초기값이 필요하여 `Result.success(listOf())` 로 설정하였으나,
`Result.success` 일 때 응답 데이터 리스트가 비어 있으므로 _"검색 결과가 없습니다"_ 메세지 표출되는 문제
-> _검색이 진행되고 있음_ 을 나타내는 상태 값(`Loading`)을 추가한 별도 클래스 생성
- 최초 검색 이후에 다시 검색할 경우, 이전 검색 상태가 `StateFlow` 에 저장되어 있는 상태이므로
새로운 검색이 진행되는 동안 이전 검색 상태가 그대로 보여지는 문제
-> 검색을 시작하면 `StateFlow` 에 다시 초기값(`Loading`)을 강제로 주입

fried-shoes committed:
feat: LiveData 를 Flow 로 변환

fried-shoes committed:
feat: 로딩 상태를 나타내기 위한 응답 결과 클래스 생성

- sealed class 만들고 Success, Error 및 Loading 추가
- MutableStateFlow 의 초기값을 Loading 으로 설정, 데이터를 받아올 때까지 로딩 화면 표출

fried-shoes committed:
feat: 다시 검색할 때마다 로딩 화면 표출

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/BookSearchRepository.kt:
```diff
@@ -1,27 +1,28 @@
-    private val _searchResult: MutableLiveData<Result<List<BookItem>>> = MutableLiveData()
-    val searchResult: LiveData<Result<List<BookItem>>> get() = _searchResult
+    private val _searchResult: MutableStateFlow<ResultUiState> = MutableStateFlow(ResultUiState.Loading)
+    val searchResult: StateFlow<ResultUiState> get() = _searchResult
```
중요한건 아닐 수 있지만, 이왕이면 StateFlow 로 변환해서 반환하는게 좋습니다.
변환은 `asStateFlow()` 함수를 사용하면 됩니다!

변환하는게 좋은 이유는 외부에서 강제로 MutableStateFlow 로 캐스팅해서 사용이 가능하기 때문입니다.
`asStateFlow()` 를 사용하면 내부적으로 `ReadonlyStateFlow` 로 감싸기 때문에 안전해집니다.

```suggestion
    val searchResult: StateFlow<ResultUiState> get() = _searchResult.asStateFlow()
```

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/data/BookSearchRepository.kt:
```diff
@@ -1,27 +1,28 @@
         try {
             val bookResponse = kakaoService.getBookList(keyword)
-            _searchResult.value = Result.success(bookResponse.bookList)
+            _searchResult.value = ResultUiState.Success(bookResponse.bookList)
```
MutableStateFlow 에는 value 에 바로 set 하는 것도 방법이지만
`update` 라는 함수도 존재합니다. 

꼭 `update` 로 변경하라는 의미의 코멘트는 아니고, 
이 둘의 동작에 대한 차이를 한번 찾아보고 학습해보면 좋을 것 같아서 남겨봅니다~!

fried-shoes replied:
#### `MutableStateFlow.value = newValue`

> 1. 새로운 값이 기존(현재) 값의 영향을 받지 않을 때
> 2. 단일 스레드 환경에서 사용
> 3. 기존 값과 비교하지 않고 새로운 상태 값을 강제 주입?

#### `MutableStateFlow.update { }`

> 1. 새로운 값이 기존(현재) 값과 관련 있거나 의존해야 하는 경우
> 2. `compareAndSet(expect, update)` : `expect` 와 `update` 을 _원자적으로 비교_ , 같으면 `update` 로 설정되어 `true` 반환
> _(원자적 비교: 메모리에 저장된 값과 현재 스레드의 값을 비교함)_
> -> 다른 스레드에서 값을 변경하지 않았을 경우 `true` 반환하여 새로운 값으로 변경됨
> 3. 스레드 안전성: 내부적으로 가장 최근 값을 비교하기 때문에 여러 스레드에서 접근하더라도 경합이 일어나지 않음

## 

#### ex) `value` or `update` 사용
```kotlin

private val _myFlow: MutableStateFlow<Int> = MutableStateFlow(0)
val myFlow: StateFlow<Int> get() = _myFlow.asStateFlow()

fun newFunction() {
    // value
    _myFlow.value++

    // update
    _myFlow.update { currentState ->
        currentState + 1
    }
}
```

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/HomeFragment.kt:
```diff
@@ -28,8 +32,12 @@ class HomeFragment: Fragment() {
-        vm.keyword.observe(viewLifecycleOwner) { keyword ->
-            binding.tlSearch.editText?.setText(keyword)
+        viewLifecycleOwner.lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
```
요거 대신 `flowWithLifecycle` 를 사용할 수도 있습니다!
(동작은 동일합니다)

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/HomeViewModel.kt:
```diff
@@ -1,15 +1,19 @@
-    private val _keyword = MutableLiveData<String>()
-    val keyword: LiveData<String> = _keyword
+    private val _keyword: MutableStateFlow<String> = MutableStateFlow("")
+    val keyword: StateFlow<String> = _keyword
```
여기도 `asStateFlow()` 를 해주면 좋을 것 같습니다.

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/ListFragment.kt:
```diff
@@ -41,27 +46,40 @@ class ListFragment: Fragment() {
-                        is HttpException -> getString(R.string.error_server)
-                        else -> getString(R.string.error_unknown)
+        viewLifecycleOwner.lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
+                vm.searchResult.collect { uiState ->
```
여기도 flowWithLifecycle 를 쓸 수 있는데 선택의 문제라 적용안해도 괜찮습니다 🙏

f-lab-tyler commented on app/src/main/java/kr/yjkim/book_search/ui/ListViewModel.kt:
```diff
@@ -1,18 +1,18 @@
 class ListViewModel(private val repository: BookSearchRepository): ViewModel() {
 
-    val searchResult: LiveData<Result<List<BookItem>>> = repository.searchResult
+    val searchResult: StateFlow<ResultUiState> = repository.searchResult
```
요기도 `asStateFlow()` 쓰면 좋을 것 같네요 😄

fried-shoes replied:
`searchResult` 의 타입이 이미 `StateFlow` 이고 `MutableStateFlow` 가 아니어서 `asStateFlow()` 가 적용이 안 되는데... 
그럼 `MutableStateFlow` 로 캐스팅한 다음 다시 `asStateFlow()` 를 써야 할까요? 

```suggestion
    val searchResult: StateFlow<ResultUiState> = (repository.searchResult as MutableStateFlow).asStateFlow()
```

f-lab-tyler replied:
아! 여긴 searchResult 가 이미 StateFlow 이군요.

그럼 괜찮습니다!

f-lab-tyler reviewed (commented):
첫 적용인데 `repeatOnLifecycle` 까지 쓴거 보니 많이 찾아보신거 같네요 👍

fried-shoes committed:
feat: MutableStateFlow 를 Read-only 설정

- MutableStateFlow.asStateFlow()
: 내부적으로 ReadOnlyStateFlow 적용

fried-shoes committed:
feat: 단순 Flow collect 의 경우 간단 작성법

- flowWithLifecycle() 의 두번째 parameter default 값이 Lifecycle.State.STARTED 이므로 생략 가능

fried-shoes commented on app/src/main/java/kr/yjkim/book_search/ui/HomeFragment.kt:
```diff
@@ -33,11 +32,10 @@ class HomeFragment: Fragment() {
         viewLifecycleOwner.lifecycleScope.launch {
-            repeatOnLifecycle(Lifecycle.State.STARTED) {
-                vm.keyword.collect { keyword ->
+            vm.keyword.flowWithLifecycle(viewLifecycleOwner.lifecycle)
```
Android developers 공식 문서 참고

> 수명 주기 인식 흐름 수집 (Lifecycle-aware flow collection)
> ---
> 단일 흐름에서 수명 주기 인식 수집을 진행하기만 하면 되는 경우 `Flow.flowWithLifecycle()` 메서드를 사용하여 코드를 단순하게 작성하면 됩니다. (중략)
> 
> _Ref. https://developer.android.com/topic/libraries/architecture/coroutines?hl=ko#restart_
